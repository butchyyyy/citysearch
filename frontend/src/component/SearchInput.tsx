import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import React from "react"
import { Input, InputGroup, InputGroupAddon, InputGroupText } from "reactstrap"

interface Props {
  /** Actual search input value */
  value: string
  /** Method handling search submit */
  onSubmitSearch: () => void
  /** Method handling search input */
  onSearchInput: (query: string) => void
}

/**
 * Component rendering user city search input p
 */
const SearchInput = (props: Props) => {
  /** Handle user input change */
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    props.onSearchInput(e.target.value)
  }

  /** Hender 'Emter' keypress input, that submits search */
  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === "Enter") {
      props.onSubmitSearch()
    }
  }

  return (
      <InputGroup>
        <InputGroupAddon addonType="prepend">
          <InputGroupText>
            <FontAwesomeIcon icon="search-location" />
          </InputGroupText>
        </InputGroupAddon>
        <Input value={props.value} onChange={handleChange} onKeyPress={handleKeyPress} placeholder="Enter City Name, Maximum of 1000 results will be shown" />
      </InputGroup>
  )
}

export default SearchInput
