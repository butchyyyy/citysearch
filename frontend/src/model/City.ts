export default interface City {
  name: string
  country: string
  population: number
  lat: number
  lng: number
}