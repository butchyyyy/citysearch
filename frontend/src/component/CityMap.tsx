import CityMarker from "component/CityMarker"
import City from "model/City"
import React from "react"
import { GoogleMap, withGoogleMap, WithGoogleMapProps, withScriptjs, WithScriptjsProps } from "react-google-maps"
import { MarkerWithLabel } from "react-google-maps/lib/components/addons/MarkerWithLabel"

interface OwnProps {
  cities: City[]
}

interface WrapperProps {
  mapApiKey: string
}

type Props = OwnProps & WithGoogleMapProps & WithScriptjsProps

class CityMap extends React.PureComponent<Props> {
  mapRef: React.RefObject<GoogleMap>

  constructor(props) {
    super(props)
    this.mapRef = React.createRef()
  }

  render() {
    return (
        <GoogleMap
            defaultCenter={{ lat: 0, lng: 0 }}
            defaultZoom={2}
            ref={this.mapRef}
        >
          {this.props.cities.map((city, index) => (
              <CityMarker key={index} city={city} />
          ))}
        </GoogleMap>
    )
  }

  componentDidUpdate() {
    if (this.props.cities && this.props.cities.length > 0) {
      const bounds = new google.maps.LatLngBounds()
      this.props.cities.forEach((city) => {
        bounds.extend({ lat: city.lat, lng: city.lng })
      })
      this.mapRef.current.fitBounds(bounds)
    }
  }
}

const Enhanced = withScriptjs(withGoogleMap(CityMap))

const composed = (props: OwnProps & WrapperProps) => (
    <Enhanced
        cities={props.cities}
        googleMapURL={`https://maps.googleapis.com/maps/api/js?key=${props.mapApiKey}`}
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `600px` }} />}
        mapElement={<div style={{ height: `100%` }} />}
    />
)

export default composed
