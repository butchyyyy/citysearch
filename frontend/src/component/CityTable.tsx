import City from "model/City"
import React from "react"
import { Table } from "reactstrap"

interface Props {
  /** Cities to tender in the table */
  cities: City[]
}

/** Renders cities in a basic responsive table */
const CityTable = (props: Props) => (
    <Table>
      <thead>
      <tr>
        <th>City</th>
        <th>Country</th>
        <th>Population</th>
        <th>Latitude</th>
        <th>Longitude</th>
      </tr>
      </thead>
      <tbody>
      {props.cities.map((city, index) => (
          <tr key={index}>
            <td>{city.name}</td>
            <td>{city.country}</td>
            <td>{city.population}</td>
            <td>{city.lat}</td>
            <td>{city.lng}</td>
          </tr>
      ))}
      </tbody>
    </Table>
)

export default CityTable
