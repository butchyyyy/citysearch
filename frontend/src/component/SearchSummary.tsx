import City from "model/City"
import React from "react"
import { Badge } from "reactstrap"

interface Props {
  /** Current search result */
  cities: City[]
}

/** Shows summary for current search result as a number of cities */
const SearchSummary = (props: Props) => (
    <h6>Cities: <Badge color="secondary">{props.cities && props.cities.length}</Badge></h6>
)

export default SearchSummary
