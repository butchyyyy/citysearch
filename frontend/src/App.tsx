import classNames from "classnames"
import CityMap from "component/CityMap"
import CityTable from "component/CityTable"
import Navbar from "component/Navbar"
import SearchInput from "component/SearchInput"
import SearchSummary from "component/SearchSummary"
import Spinner from "component/Spinner"
import City from "model/City"
import React from "react"
import { Col, Container, CustomInput, Nav, NavItem, NavLink, Row, TabContent, TabPane } from "reactstrap"

interface AppProps {
  mapApiKey: string
}

interface State {
  activeTab: string
  searchInput: string
  searchResult: City[]
  searching: boolean
  markerClusterer: boolean
}

class App extends React.Component<AppProps, State> {
  constructor(props) {
    super(props)
    this.state = {
      searching: false,
      activeTab: "1",
      searchInput: "",
      searchResult: [],
      markerClusterer: true,
    }
    this.toggle = this.toggle.bind(this)
    this.toggleTab1 = this.toggleTab1.bind(this)
    this.toggleTab2 = this.toggleTab2.bind(this)
    this.searchInput = this.searchInput.bind(this)
    this.submitSearch = this.submitSearch.bind(this)
    this.markerClustererToggle = this.markerClustererToggle.bind(this)
  }

  toggle(tab) {
    if (this.state.activeTab !== tab) {
      this.setState({ activeTab: tab })
    }
  }

  toggleTab1() {
    this.toggle("1")
  }

  toggleTab2() {
    this.toggle("2")
  }

  searchInput(query: string) {
    this.setState({ searchInput: query })
  }

  submitSearch() {
    this.setState({ searching: true })
    return fetch(`/cities/search?query=${this.state.searchInput}`)
        .then((response: Response) => {
          if (response.ok) {
            response.json().then((data) => this.setState({ searching: false, searchResult: data }))
          } else {
            this.setState({ searching: false, searchResult: [] })
            window.alert("Failed to load cities: " + response.statusText)
          }
        })
        .catch((error) => {
          this.setState({ searching: false, searchResult: [] })
          window.alert("Failed to load cities: " + error)
        })
  }

  markerClustererToggle() {
    this.setState({ markerClusterer: !this.state.markerClusterer })
  }

  render() {
    return (
        <Container>
          <Navbar />
          <Spinner loading={this.state.searching}>
            <Row>
              <Col>
                <SearchInput value={this.state.searchInput} onSearchInput={this.searchInput} onSubmitSearch={this.submitSearch} />
              </Col>
            </Row>
            <Row className="row-top-buffer-small">
              <Col xs={6} sm={6} md={4} lg={4} xl={2}>
                <SearchSummary cities={this.state.searchResult} />
              </Col>
              <Col xs={6} sm={6} md={4} lg={4} xl={2}>
                <CustomInput
                    type="switch"
                    label="Cluster Markers"
                    id="markerClusetere"
                    checked={this.state.markerClusterer}
                    onChange={this.markerClustererToggle}
                />
              </Col>
            </Row>
            <Row className="row-top-buffer-small">
              <Col>
                <Nav tabs>
                  <NavItem>
                    <NavLink
                        className={classNames({ active: this.state.activeTab === "1" })}
                        onClick={this.toggleTab1}
                    >
                      Map View
                    </NavLink>
                  </NavItem>
                  <NavItem>
                    <NavLink
                        className={classNames({ active: this.state.activeTab === "2" })}
                        onClick={this.toggleTab2}
                    >
                      Table view
                    </NavLink>
                  </NavItem>
                </Nav>
                <TabContent activeTab={this.state.activeTab}>
                  <TabPane tabId="1">
                    <Row className="row-top-buffer">
                      <Col>
                        <CityMap mapApiKey={this.props.mapApiKey} cities={this.state.searchResult} markerClusterer={this.state.markerClusterer} />
                      </Col>
                    </Row>
                  </TabPane>
                  <TabPane tabId="2">
                    <Row className="row-top-buffer">
                      <Col>
                        <CityTable cities={this.state.searchResult} />
                      </Col>
                    </Row>
                  </TabPane>
                </TabContent>
              </Col>
            </Row>
          </Spinner>
        </Container>
    )
  }
}

export default App
