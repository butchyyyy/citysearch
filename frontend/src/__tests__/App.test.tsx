import { testCities } from "__tests__/TestData"
import App from "App"
import CityMap from "component/CityMap"
import CityTable from "component/CityTable"
import SearchInput from "component/SearchInput"
import SearchSummary from "component/SearchSummary"
import { shallow } from "enzyme"
import "jest-fetch-mock"
import City from "model/City"
import React from "react"
import { CustomInput, NavLink } from "reactstrap"

describe("Test App Component", () => {
  const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => null)
  const wrapper = shallow<App>(<App mapApiKey="" />)

  const testCityComponentProp = (expectedValue: City[]) => {
    expect(wrapper.find(SearchSummary).props().cities).toEqual(expectedValue)
    expect(wrapper.find(CityMap).props().cities).toEqual(expectedValue)
    expect(wrapper.find(CityTable).props().cities).toEqual(expectedValue)
  }

  it("Test App Snapshost", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("Test Fetch Data", () => {
    testCityComponentProp([])
    fetchMock.mockResponseOnce(JSON.stringify(testCities))
    return wrapper.instance().submitSearch().then(() => {
      testCityComponentProp(testCities)
    })
  })

  it("Test Fetch Data Not Ok Response", () => {
    fetchMock.mockResponseOnce(null, { status: 400, statusText: "Bad Request" })
    return wrapper.instance().submitSearch().then(() => {
      testCityComponentProp([])
      expect(alertSpy).toHaveBeenCalledWith("Failed to load cities: Bad Request")
    })
  })

  it("Test Fetch Data Rejectd", () => {
    fetchMock.mockRejectOnce(new Error("Test Error"))
    return wrapper.instance().submitSearch().catch(() => {
      testCityComponentProp([])
      expect(alertSpy).toHaveBeenCalledWith("Failed to load cities: Test Error}")
    })
  })

  it("Test Search Input", () => {
    expect(wrapper.find(SearchInput).props().value).toEqual("")
    wrapper.instance().searchInput("foo")
    expect(wrapper.find(SearchInput).props().value).toEqual("foo")
  })

  it("Test Tab Toggle", () => {
    expect(wrapper.find(NavLink).at(0).hasClass("active")).toBeTruthy()
    expect(wrapper.find(NavLink).at(1).hasClass("active")).toBeFalsy()
    wrapper.find(NavLink).at(1).simulate("click")
    expect(wrapper.find(NavLink).at(0).hasClass("active")).toBeFalsy()
    expect(wrapper.find(NavLink).at(1).hasClass("active")).toBeTruthy()
    wrapper.find(NavLink).at(0).simulate("click")
    expect(wrapper.find(NavLink).at(0).hasClass("active")).toBeTruthy()
    expect(wrapper.find(NavLink).at(1).hasClass("active")).toBeFalsy()
  })

  it("Test Cluster Marker Toggle", () => {
    expect(wrapper.find(CityMap).props().markerClusterer).toBeTruthy()
    const markerClusterSwitch = wrapper.find(CustomInput)
    markerClusterSwitch.simulate("change")
    expect(wrapper.find(CityMap).props().markerClusterer).toBeFalsy()
    markerClusterSwitch.simulate("change")
    expect(wrapper.find(CityMap).props().markerClusterer).toBeTruthy()
  })
})
