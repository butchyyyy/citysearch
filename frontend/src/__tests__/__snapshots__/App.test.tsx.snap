// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App Component Test App Snapshost 1`] = `
<Container
  tag="div"
>
  <Navbar />
  <Spinner
    loading={false}
  >
    <Row
      tag="div"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <SearchInput
          onSearchInput={[Function]}
          onSubmitSearch={[Function]}
          value=""
        />
      </Col>
    </Row>
    <Row
      className="row-top-buffer-small"
      tag="div"
    >
      <Col
        lg={4}
        md={4}
        sm={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xl={2}
        xs={6}
      >
        <SearchSummary
          cities={Array []}
        />
      </Col>
      <Col
        lg={4}
        md={4}
        sm={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xl={2}
        xs={6}
      >
        <CustomInput
          checked={true}
          id="markerClusetere"
          label="Cluster Markers"
          onChange={[Function]}
          type="switch"
        />
      </Col>
    </Row>
    <Row
      className="row-top-buffer-small"
      tag="div"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Nav
          tabs={true}
          tag="ul"
          vertical={false}
        >
          <NavItem
            tag="li"
          >
            <NavLink
              className="active"
              onClick={[Function]}
              tag="a"
            >
              Map View
            </NavLink>
          </NavItem>
          <NavItem
            tag="li"
          >
            <NavLink
              className=""
              onClick={[Function]}
              tag="a"
            >
              Table view
            </NavLink>
          </NavItem>
        </Nav>
        <TabContent
          activeTab="1"
          tag="div"
        >
          <TabPane
            tabId="1"
            tag="div"
          >
            <Row
              className="row-top-buffer"
              tag="div"
            >
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <composed
                  cities={Array []}
                  mapApiKey=""
                  markerClusterer={true}
                />
              </Col>
            </Row>
          </TabPane>
          <TabPane
            tabId="2"
            tag="div"
          >
            <Row
              className="row-top-buffer"
              tag="div"
            >
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <CityTable
                  cities={Array []}
                />
              </Col>
            </Row>
          </TabPane>
        </TabContent>
      </Col>
    </Row>
  </Spinner>
</Container>
`;
